### DISCUSSION TITLE
Changing MTU Size

### DISCUSSION THREAD

**Discussion by mwswt** [shinyorg/shiny#1213] [Q&A] [answered]
Hello Shiny.NET Team,

i find your library for .NET MAUI really great and it is working fine so far. But i have aproblem with the MTU size. I have looked at the code and it seems so that the MTU size can not be changed at the moment. Is it planned to implement this feature?

Thank you in advance for your answers.

Best regards,

Markus

**Comment by aritchie** [comment]
It really isn't a team... it's just me :)

MTU is now baked in with the connection request since it only applies to android, ios negotiates this upon connection for you.

To request MTU on android

After discovering your peripheral
```cs
IPeripheral p;

#if ANDROID
peripheral.Connect(new AndroidConnectionConfig(
    RequestedMtu = 250
));
#endif
```

as seen here:
https://github.com/shinyorg/shiny/blob/master/src/Shiny.BluetoothLE/Platforms/Android/AndroidConnectionConfig.cs
https://github.com/shinyorg/shiny/blob/45239d9958ce6164e57935c9dc84fba123b8be4c/src/Shiny.BluetoothLE/Platforms/Android/Peripheral.cs#L76

**Reply by mwswt** [reply]
Hello Alan,

thank you for your answer, but i don't get this to work. I see in the LOG that the MTU size is changed, but the callback function OnMTUChanged will not be called.
Is there something i made wrong?

Here is a snippet from my log:
[BluetoothAdapter] STATE_ON
[BluetoothGatt] connect() - device: 24:6F:28:F6:11:DD, auto: true
[BluetoothAdapter] isSecureModeEnabled
[BluetoothGatt] registerApp()
[BluetoothGatt] registerApp() - UUID=6cd0d9d6-9bea-4eae-b105-a5c9a6152bb4
[BluetoothGatt] onClientRegistered() - status=0 clientIf=13
[BluetoothGatt] requestConnectionPriority() - params: 0
[BluetoothGatt] configureMTU() - device: 24:6F:28:F6:11:DD mtu: 512
[name.Playgroun] Explicit concurrent copying GC freed 2391(255KB) AllocSpace objects, 0(0B) LOS objects, 57% free, 4539KB/10MB, paused 22us,20us total 14.966ms
[GestureDetector] handleMessage TAP
[name.Playgroun] Compiler allocated 4163KB to compile void android.widget.TextView.<init>(android.content.Context, android.util.AttributeSet, int, int)
[BluetoothGatt] onClientConnectionState() - status=0 clientIf=13 device=24:6F:28:F6:11:DD
Thread started:  #16

Best regards,

Markus

**Reply by aritchie** [reply]
It just isn't updating the MTU value on the peripheral code, but the MTU request is clearly working so you haven't done anything wrong

**Comment by mwswt** [comment]
Hello Alan,

thank you for the information. I have another question.
Is this a bug and will it be fixed?
So what i have tried, i made a break point after the connect in the Peripheral (Android) and wait a little then i have clicked on Run again and then the MTU was updated in the Peripheral.
For me it looks like a timing problem between connect and change mtu.

Best regards,

Markus

**Comment by jaredballen** [comment]
@mwswt, the `ICanRequestMtu` interface has been added to the `Peripheral` class. After you have successfully connected to your peripheral and before you start service discovery, you can use pattern matching on the peripheral instance for this interface and request the desired MTU size. The `ICanRequestMtu` is only implemented on Android.

**Reply by aritchie** [reply, marked as answer]
There is a better extension you can use called TryMtuRequest(this IPerihperal peripheral, int mtu) that you can use.  If MTU requests are available, it will do it otherwise it will move on.  It will also return the actual negotiated value which often isn't the value you request.

**Comment by powerteqtpitman** [comment]
I am having an issue related to this and hope it is ok if I jump into this discussion to ask about it.

I have an app that I am upgrading from Xamarin to MAUI. I started a new MAUI project and am porting over the code. I have done this a few times now with success.

I have 2 different BLE devices that I am working with between 2 different apps.

I know that TryRequestMtu is only provided on Android because iOS auto negotiates it.

What is weird is the following observations:

1. When my app was a Xamarin app, and I called TryRequestMtu(129) that method would return 129 on iOS and Android.
2. When I now call TryRequestMtu(129) on the same app as a MAUI app and the same device, it returns 20 on iOS and 129 on Android.
3. When I use the MAUI version same app in #2 to connect to my other device (totally different BLE chip), it returns 129 on iOS and Android.

The device that shows the MTU I am asking for (129) on both is based on the Nordic chip. The device that used to show 129 on iOS and now shows 20 on iOS is based on the esp32.

EDIT: I forgot to mention that I am currently using the latest 3.3.3 version of all Shiny nuget packages and net8.0.

Thoughts?

**Reply by powerteqtpitman** [reply]
I kept playing around with this and figured out what the issue is.

I was calling TryRequestMtu method right after getting the Connected state change from Shiny.

I added a simple 2 second delay and now the mtu is coming back as 253, which is what the device supports.

Sorry to highjack this thread. Maybe if someone else has a similar issue this will help them.

I used 2 seconds just to see if a delay would help. After that worked I was able to reduce it down to 90ms before it quit updating fast enough. I changed it to 100ms just to have a small margin.
